lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -0,0 +1,712 @@
+                        debug!("All channels closed, exiting dispatcher");
+                        break;
+                    }
+                }
+            }
+            debug!("Check dispatcher shutdown complete");
+        });
+
+        check_metrics_rx
+    }
+}
+
+#[derive(Clone)]
+pub struct ChecksTelemetry {
+    check_requests_dispatched: Counter,
+    check_instances_started: Counter,
+}
+
+#[cfg(test)]
+impl ChecksTelemetry {
+    fn noop() -> Self {
+        Self {
+            check_requests_dispatched: Counter::noop(),
+            check_instances_started: Counter::noop(),
+        }
+    }
+}
+
+pub struct Checks {
+    listeners: Vec<listener::DirCheckRequestListener>,
+}
+
+impl Checks {
+    // consumes self
+    async fn run_inner(self, context: SourceContext, global_shutdown: ComponentShutdownHandle) -> Result<(), ()> {
+        let mut listener_shutdown_coordinator = DynamicShutdownCoordinator::default();
+        let metrics = MetricsBuilder::from_component_context(context.component_context());
+        let tlm = ChecksTelemetry {
+            check_requests_dispatched: metrics.register_debug_counter("checkrequests.dispatched"),
+            check_instances_started: metrics.register_debug_counter("checkinstances.started"),
+        };
+
+        let Checks { listeners } = self;
+
+        let (check_run_requests_tx, check_run_requests_rx) = mpsc::channel(100);
+        let (check_stop_requests_tx, check_stop_requests_rx) = mpsc::channel(100);
+        let dispatcher =
+            CheckDispatcher::new(check_run_requests_rx, check_stop_requests_rx, tlm).expect("Could create");
+
+        let mut joinset = tokio::task::JoinSet::new();
+        // Run the local task set.
+        // For each listener, spawn a dedicated task to run it.
+        for listener in listeners {
+            let listener_context = listener::DirCheckListenerContext {
+                shutdown_handle: listener_shutdown_coordinator.register(),
+                submit_runnable_check_req: check_run_requests_tx.clone(),
+                submit_stop_check_req: check_stop_requests_tx.clone(),
+                listener,
+            };
+
+            joinset.spawn(process_listener(context.clone(), listener_context));
+        }
+
+        info!("Check source started.");
+
+        let mut check_metrics_rx = dispatcher.run();
+
+        let check_shutdown = listener_shutdown_coordinator.register();
+        tokio::spawn(async {
+            global_shutdown.await;
+            listener_shutdown_coordinator.shutdown().await;
+        });
+
+        tokio::pin!(check_shutdown);
+
+        loop {
+            select! {
+                Some(check_metric) = check_metrics_rx.recv() => {
+                    info!("Received check metric: {:?}", check_metric);
+                    let mut event_buffer = context.event_buffer_pool().acquire().await;
+                    let event: Event = check_metric.try_into().expect("can't convert");
```

Comment:
I believe this is a starting point for the breadcrumb of "what metric types get translated to what". This `try_into` `Event` is doing a lot of work.